core/src/main/java/org/apache/iceberg/hadoop/Util.java

```diff
@@ -84,10 +84,16 @@ public static String[] blockLocations(FileIO io, ScanTaskGroup<?> taskGroup) {
     return locations.toArray(HadoopInputFile.NO_LOCATION_PREFERENCE);
   }
 
+  public static boolean isHDFSLocation(FileIO io, String location) {
+    return io instanceof HadoopFileIO
+        || (io instanceof ResolvingFileIO
+            && ResolvingFileIO.implFromLocation(location).equals(HadoopFileIO.class.getName()));
+  }
+
   private static String[] blockLocations(FileIO io, ContentScanTask<?> task) {
-    InputFile inputFile = io.newInputFile(task.file().path().toString());
-    if (inputFile instanceof HadoopInputFile) {
-      HadoopInputFile hadoopInputFile = (HadoopInputFile) inputFile;
+    String location = task.file().path().toString();
+    if (isHDFSLocation(io, location)) {
+      HadoopInputFile hadoopInputFile = (HadoopInputFile) io.newInputFile(location);
```

Comment:
If we plan to support custom `FileIO` implementations that extend `HadoopFileIO`, this cast may be not safe.

Reply:
ACk made the changes